cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -91,51 +201,730 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+   * The {@link StreamSizeScheduler} communicates with this class, which in turn communicates to the schedules
+   * it contains to perform operations on the schedules - suspend, resume, etc.
+   */
+  private final class StreamSubscriber implements NotificationHandler<StreamSizeNotification>, Cancellable {
+    // Key is the schedule ID
+    private final ConcurrentMap<String, StreamSizeScheduleTask> scheduleTasks;
+    private final Id.Stream streamId;
+    private final AtomicInteger activeTasks;
+
+    private Cancellable notificationSubscription;
+    private ScheduledFuture<?> scheduledPolling;
+    private StreamSizeNotification lastNotification;
+    private StreamSizeNotification lastPollingInfo;
+
+    private StreamSubscriber(Id.Stream streamId) {
+      this.streamId = streamId;
+      this.scheduleTasks = Maps.newConcurrentMap();
+      this.activeTasks = new AtomicInteger(0);
+    }
+
+    public void start() throws NotificationFeedException, NotificationFeedNotFoundException {
+      notificationSubscription = notificationService.subscribe(getFeed(), this, sendPollingInfoExecutor);
+    }
+
+    @Override
+    public void cancel() {
+      if (scheduledPolling != null) {
+        scheduledPolling.cancel(true);
+      }
+      if (notificationSubscription != null) {
+        notificationSubscription.cancel();
+      }
+    }
+
+    /**
+     * Add a new scheduling task based on the data received by the stream referenced by {@code this} object.
+     * @return {@code true} if the task was created successfully, {@code false} if it already exists
+     */
+    public boolean createScheduleTask(Id.Program programId, SchedulableProgramType programType,
+                                      StreamSizeSchedule streamSizeSchedule, boolean active,
+                                      long baseRunSize, long baseRunTs, boolean persist) throws SchedulerException {
+      final StreamSizeScheduleTask newTask = new StreamSizeScheduleTask(programId, programType, streamSizeSchedule);
+      synchronized (this) {
+        StreamSizeScheduleTask previous =
+          scheduleTasks.putIfAbsent(AbstractSchedulerService.scheduleIdFor(programId, programType,
+                                                                           streamSizeSchedule.getName()),
+                                    newTask);
+        if (previous != null) {
+          // We cannot replace an existing schedule - that functionality is not wanted - yet
+          return false;
+        }
+
+        if (active) {
+          activeTasks.incrementAndGet();
+        }
+      }
+
+      // Initialize the schedule task
+      if (baseRunSize == -1 && baseRunTs == -1) {
+        // This is the first time that we schedule this task - ie it was not in the schedule store
+        // before. Hence we set the base info properly by polling the stream
+        try {
+          StreamSize streamSize = queryStreamEventsSize();
```

Comment:
these two calls seem to come together almost all the time. I feel that the code could be simplified significantly if you restructure it. Let's talk in person